PhiloxRandom: Fix race in GPU fill function

jthestness opened the pull request:
The PhiloxRandom fill kernel for the GPU had race conditions that caused the
outputs to be non-deterministic. In particular, the code previously executed
with N GPU threads (# thread contexts per GPU), but it would only advance the
fill addresses by N-1 stride in each step. This incorrect stride caused the
0th and N-1st threads to write to the same memory locations, racing for which
was last to write their common locations. Make the stride equal to the number
of threads to eliminate the race.

BONUS: By fixing this race, PhiloxRandom constant-sized GPU initializers now
match CPU initializers.

jthestness committed:
PhiloxRandom: Fix race in GPU fill function

The PhiloxRandom fill kernel for the GPU had race conditions that caused the
outputs to be non-deterministic. In particular, the code previously executed
with N GPU threads (# thread contexts per GPU), but it would only advance the
fill addresses by N-1 stride in each step. This incorrect stride caused the
0th and N-1st threads to write to the same memory locations, racing for which
was last to write their common locations. Make the stride equal to the number
of threads to eliminate the race.

BONUS: By fixing this race, PhiloxRandom constant-sized GPU initializers now
match CPU initializers.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

ekelsen commented on tensorflow/core/kernels/random_op_gpu.cu.cc:
```diff
@@ -141,7 +141,7 @@ struct FillPhiloxRandomKernel<Distribution, false> {
 
-      offset += (total_thread_count - 1) * kGroupSize;
+      offset += total_thread_count * kGroupSize;
       gen.Skip(total_thread_count - 1);
```
I think you also want to skip total_thread_count as well.

jthestness replied:
I thought so too, but when I changed this, it gave different results than the CPU version of the initializer. Somehow, the generator does not generate repeated values even though this code appears to only skip `total_thread_count - 1` indices in the randomized sequence (i.e. seems like we should expect that the 0th and `total_thread_count - 1`st threads would produce the same results in successive steps). Perhaps there is an off-by-one issue inside the generator code that is consistent between CPU and GPU versions?

Do you have thoughts/recommendations for how to proceed here?

ekelsen replied:
Hmm...you've verified that the 0th thread on the 2nd pass through the loop doesn't generate the same result as the last thread on the first pass?  It seems like that must be the case.  I'd like to understand how that's possible.

jthestness replied:
Yes, I verified that. The PhiloxRandom::Skip function appears correct, so I'm also not sure how there can be an off-by-one in there.

ekelsen replied:
Can you examine the state of the kernel for those two threads using cuda-gdb or printfs?  I'd like to make sure we understand what is actually going on - there's probably something we're missing.

jthestness replied:
Hah! Silly: The generator automatically advances one position when it is read. Thus, when you skip after that, you only need to skip `total_thread_count - 1`

ekelsen commented:
Would be great if you could also add a test that would fail before your change and works after.

jthestness committed:
Update random_ops_test.py to find race conditions

Increasing the size of arrays in the random_ops_test.py test to manifest
the race conditions to be resolved.

jthestness commented:
@ekelsen: I updated the existing `random_ops_test.py` to manifest the race condition

ekelsen commented:
Haha, good catch!  LGTM

jhseu commented:
Jenkins, test this please
